src/lib.rs

```diff
@@ -1,334 +1,201 @@
-            .arg(&desc_file);
-        for f in files {
-            cmd.arg(&format!("{}", f.as_ref().display()));
-        }
-        println!("executing {:?}", cmd);
-        match cmd.status() {
-            Ok(e) if e.success() => {}
-            e => panic!("failed to generate descriptor set files: {:?}", e),
-        }
 
-        let desc_bytes = std::fs::read(&desc_file).unwrap();
-        let desc: protobuf::descriptor::FileDescriptorSet =
-            protobuf::parse_from_bytes(&desc_bytes).unwrap();
-        let mut files_to_generate = Vec::new();
-        'outer: for file in files {
-            for include in includes {
-                if let Ok(truncated) = file.as_ref().strip_prefix(include) {
-                    files_to_generate.push(format!("{}", truncated.display()));
-                    continue 'outer;
+    /// Finds proto files to operate on in the `proto_dir` directory.
+    pub fn search_dir_for_protos(&mut self, proto_dir: &str) -> &mut Self {
+        self.files = fs::read_dir(proto_dir)
+            .expect("Couldn't read proto directory")
+            .filter_map(|e| {
+                let e = e.expect("Couldn't list file");
+                if e.file_type().expect("File broken").is_dir() {
+                    None
+                } else {
+                    Some(format!("{}/{}", proto_dir, e.file_name().to_string_lossy()))
                 }
-            }
-
-            panic!("file {:?} is not found in includes {:?}", file, includes);
-        }
-
-        protobuf_codegen::gen_and_write(
-            desc.get_file(),
-            &files_to_generate,
-            &Path::new(out_dir),
-            &protobuf_codegen::Customize::default(),
-        )
-        .unwrap();
-        generate_grpcio(&desc.get_file(), &files_to_generate, out_dir);
-        replace_read_unknown_fields(out_dir);
+            })
+            .collect();
+        self
     }
 
-    #[cfg(feature = "grpcio-protobuf-codec")]
-    pub fn generate_grpcio(
-        desc: &[protobuf::descriptor::FileDescriptorProto],
-        files_to_generate: &[String],
-        out_dir: &str,
-    ) {
-        use std::io::Write;
-
-        let output_dir = std::path::Path::new(out_dir);
-        let results = grpcio_compiler::codegen::gen(desc, &files_to_generate);
-        for res in results {
-            let out_file = output_dir.join(&res.name);
-            let mut f = std::fs::File::create(&out_file).unwrap();
-            f.write_all(&res.content).unwrap();
-        }
+    pub fn files<T: Into<String> + Clone>(&mut self, files: &[T]) -> &mut Self {
+        self.files = files
+            .iter()
+            .map(|t| t.clone().into())
+            .collect::<Vec<String>>();
+        self
     }
 
-    #[cfg(all(feature = "protobuf-codec", not(feature = "grpcio-protobuf-codec")))]
-    pub fn generate_grpcio(_: &[protobuf::descriptor::FileDescriptorProto], _: &[String], _: &str) {
+    pub fn includes(&mut self, includes: Vec<String>) -> &mut Self {
```

Comment:
I think rust style prefer using `set_includes` instead?

Reply:
The preferred style for builders is not to use `set_`, see e.g., https://doc.rust-lang.org/nightly/std/process/struct.Command.html. I don't know why there is a difference between regular objects and builders. I'm happy to use `set_` if that is what we usually do for builders in TiKV.